pkg/pillar/docs/failover.md

```diff
@@ -0,0 +1,181 @@
+The qcow2/raw files are converted into PVCs by the volume manager and hence content tree config also has designated node id.
+This is very important for the non-container format content tree because the content is replicated to all nodes in the cluster and hence it is not necessary to download content tree to all nodes of the cluster.
+
+The container image format content tree is processed in two different ways depending on the application type.
+
+ 1) If the application type is a Container (ie container launched in shim VM), only designated node will download the content tree, rest of the nodes will get the content through replication of PVC.
+ 2) If the application type is Native container (ie NOHYPER virtualization mode), then the content tree is downloaded to all the nodes in the cluster. That is because since the container image does not contain the kernel or OS components they need to be launched natively on the kubernetes infrastructure, converting them to PVC is useless in such cases.
+
+ eve-api has been enhanced to add designated node id to content tree struct
+ Please see eve-api/proto/config/storage.proto
+ message ContentTree {
+   ....
+  // To inform the edge-node if the device receiving this content tree is
+  // responsible for content tree download or not.
+  string designated_node_id = 12;
+ }
+
+EVE contenttreetypes.go has been enhanced to include IsLocal boolean. A content tree is set to IsLocal=true if the content tree is downloaded to that node, else it is set to false. For single node setups IsLocal is always true. For native containers IsLocal is always true.
+
+Please see pkg/pillar/types/contenttreetypes.go
+type ContentTreeConfig struct {
+   ....
+   // Do we download on this node
+    IsLocal bool
+}
+type ContentTreeStatus struct {
+   ....
+   IsLocal bool
+}
+
+Zedagent microservice parses the config from the device and updates the IsLocal field in the ContentTreeConfig struct.
+
+If the ContentTree is in qcow2/raw, or container format (not native container), it gets converted to Kubernetes Persistent volume (PV) and gets replicated to all other nodes in the cluster.
+
+### Application instance
+
+Application deployed in a cluster setup will also have designated node id. That node is picked by the controller and how it picks the designated node is beyond scope of EVE. Once the designated node id is picked, controller sends Application instance config to all the devices in the cluster with designated node id set to uuid of the device which is supposed to be designated node id for that application.
+
+The designated device will start the application and publishes AppInstanceStatus. All the nodes in the cluster will publish the AppInstanceStatus since they all receive the config. But only the node that is running the app will have aiStatus.Activated is set to true. That will be used by zedagent to update the info message to the controller.
```

Comment:
I think zedagent should report the info once k3s is trying to run the app instance on the node and until it is trying to run it on a different node. Otherwise, if the user requests it to stop or restart it will not be reported by any node until it is back up and running. Can we tell that k3s has rescheduled the app instance to run on the local node?

Reply:
I am not entirely clear on this question, may be I misunderstood. 
zedagent  on designated node reports the app status like it does today on single node while deploying the app. Like downloading, loading, booting, online etc. Only the stand by nodes will not report that. 

So if user stops and starts  the app that does not trigger failover, the behavior is same as single node stop and start. Basically when user explicitly stops the app, domainmgr will delete the replica set, and hence Kubernetes will not failover the app (nothing to failover). When user starts the app it goes through normal app start process and the states are reflected accordingly in the UI.

Reply:
After the designated node has failed and the app instance has failed over to a different node I'd expect to be able to do the same operations (such as restart or stop the app instance from the controller) and see the same output (such as the info reporting the app instance halting then halted then booting if it was a restart). I'm not sure the check for "activated" ensures that this reporting happens when I stop it from the controller.

Reply:
> After the designated node has failed and the app instance has failed over to a different node I'd expect to be able to do the same operations (such as restart or stop the app instance from the controller) and see the same output (such as the info reporting the app instance halting then halted then booting if it was a restart). I'm not sure the check for "activated" ensures that this reporting happens when I stop it from the controller.

Yes I tested this scenario and it works as you are expecting.  I rebooted the designated node and the app failed over to some other node. Then I deactivated and activated app and observed that the app goes to halted state and then comes back up online. Basically was able to see all state transitions of app on non-designated node.

Reply:
Ok